### PR TITLE
Unification of DockerHub and Openshift EE image

### PR DESCRIPTION
**Pull request by hasancelik**
Maintining two different repos/Dockerfiles for almost identical EE images does not make sense. Only disadvantage is increase of image size:

Screen Shot 2020-12-18 at 12 20 28

Related PRD --> https://hazelcast.atlassian.net/wiki/spaces/PM/pages/2677604456/Unification+of+RedHat+Connect+and+DockerHub+IMDG+Enterprise+Images

At the end of the process, we can remove [hazelcast/hazelcast-openshift](https://github.com/hazelcast/hazelcast-openshift) repo, create images for RedHat registry from [hazelcast/hazelcast-docker](https://github.com/hazelcast/hazelcast-docker)

I have created [temp image](https://hub.docker.com/layers/hasancelik/hazelcast-enterprise/ubi8-minimal/images/sha256-37ca72a97d8a4e13abafeff6b282073c8b33de875d505dd5a5cdee0ce9fd30e8?context=explore) for test purposes.

fixes https://github.com/hazelcast/hazelcast-openshift/issues/39

### CODE REVIEW

File: .travis.yml
Comment (leszko) [inline .travis.yml:21]: One drawback of this unification would be that you need to have access to Red Hat Registry if you want to build the image.

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:7]: Have you tested if OpenSSL works correctly with the updated versions?

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:45]: What is this `version`? I know we had it in `hazelcast-openshift`, but is it still mandatory to have?

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:48]:
```suggestion
      summary="Hazelcast IMDG Enterprise Image, certified to RHEL" \
```

To reduce the maintainance burden.

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:49]: Could you update this description? This image is not only for OpenShift/Kubernetes.

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:29]: How about making these env similar to `hazelcast-oss`? Like not to specify `10001`?

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:67]: Do we plan to have the Eureka plugin in? I see you removed it from here but didn't remove `ARG EUREKA_PLUGIN_URLS`. I'm fine with removing from Hazelcast Enterprise image, but could you please make sure none of our customers uses Eureka atm?

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:58]: Do we need this `HZ_CP_MOUNT`? I'd keep the directories structure similar to `hazelcast-oss`.

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:69]: Could you make this part (responsible for adding new user) similar to what we have in `hazelcast-oss`?

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:55]: We still need pardot ID I guess.

**Review by leszko — COMMENTED** [review]
Added my comments. I think this change is very good in terms of maintenance. However, we need to take a careful approach. See my comments inline and below.

From the technical part, I think you could make the Dockerfile as similar to `hazlecast-oss` (or former `hazelcast-enterprise`) as possible. 

There is also the business-technical part of this. There are some non-backward-compatible implications of this change and I'd suggest creating a TDD describing them.

1. We remove Eureka plugin, we need to make sure none of our customers uses it.
2. The image size is bigger
3. The base image is Red Hat based. Technically, it should be fine, but are there any legal restrictions? Or the license for UBI is the same open as Alpine?
4. You need access to the Red Hat registry in order to build the image. Are any of our customers who build their own images and may have issues with that?
5. What are tools do for their enterprise images (like Redis, Mongo, etc.). Do they all use UBI images?

The other aspect is the performance, but we actually never checked it for the Alpine-based image, so I guess we can skip this for now.

So, I think we need a TDD and I'd like to get the review from at least @jerrinot and @huseyinbabal (Cloud uses our current Docker Hub Enterprise image).

File: hazelcast-enterprise/Dockerfile
Comment (kwart) [reply hazelcast-enterprise/Dockerfile:7]: If we are updating the lib, we should switch to even newer version `4.1.53.Final` and the latest tcnative: `2.0.34.Final`. They are currently used in the hazelcast-enterprise `master` branch.

**Review by kwart — COMMENTED** [review]
I think the unification of enterprise images (EE+openshift) could be a good step, but we need to handle it carefully.

File: .travis.yml
Comment (hasancelik) [reply .travis.yml:21]: I have changed base image from `registry.redhat.io/ubi8/ubi-minimal` to `registry.access.redhat.com/ubi8/ubi-minimal` at the recommendation of @kwart so creds are not required anymore.

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:7]: All Netty dependencies were already removed at master branch so all of them will be gone when the PR rebased.

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:45]: Actually, we used it wrong all the time, it should be image/hazelcast version. And yes it is mandatory, there are tests for it and other labels at `Certification Build`. Here is the related section from [official doc](https://connect.redhat.com/zones/containers/container-certification-policy-guide#version_label_exists):
Screen Shot 2020-12-16 at 17 28 37

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:48]: Done 👍

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:49]: Done 👍

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:29]: Done 👍

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:67]: Actually, I didn't have such a plan for this PR, it just a copy paste error. I want to remove it with another PR but I think we  can do it now 👍 I will talk with support team.

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:58]: 👍

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:69]: Done 👍

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:1]: both of them are done 👍

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile:64]: > Headless JRE should be sufficient (java-11-openjdk-headless).

good point 👍 

> If we keep the netty and tcnative libs in the image we should also install the openssl and apr packages to make it work.

netty and tcnative jars were already removed at master branch.

**Comment by pivovarit** [comment]
@hasancelik what's the updated size after the layers got merged?

**Comment by hasancelik** [comment]
@pivovarit current image size is [129 MB](https://hub.docker.com/layers/hasancelik/hazelcast-enterprise/ubi8-minimal-compress/images/sha256-37ca72a97d8a4e13abafeff6b282073c8b33de875d505dd5a5cdee0ce9fd30e8?context=explore).

**Comment by pivovarit** [comment]
Well, that's nice. 180MB did not sound good

**Comment by hasancelik** [comment]
180 MB also contains eureka jars but yes, current size is better.

**Comment by hasancelik** [comment]
@huseyinbabal and @jerrinot could you review the PR? Here is the [TDD](https://hazelcast.atlassian.net/wiki/spaces/PM/pages/2918645770/Unification+of+RedHat+Connect+and+DockerHub+IMDG+Enterprise+Images+TDD), you can find details about UBI redistribution and licensing at there.

**Review by kwart — APPROVED** [review]
LGTM :+1:

**Review by jerrinot — APPROVED** [review]
I did not do technical review. I agree with the unification and using UBI. even if it means a somewhat bigger image.

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:35]: How about changing the summary and description to just: "Hazelcast IMDG Enterprise Image"?

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:1]: Shouldn't we have the specific version of the base image here? Otherwise, we risk the following issues:
- If RH pushes by mistake the wrong `latest` image, we have an issue
- Two same `docker build` can result in different results (if `latest` is changed in the meantime)
- Some of the commands below may stop working and some point and we discover it the hard way.

File: hazelcast-enterprise/Dockerfile
Comment (leszko) [inline hazelcast-enterprise/Dockerfile:20]: Inline `USER_NAME` to move to `ARG`. I guess you don't need it at runtime.

**Review by leszko — COMMENTED** [review]
Added my comments.

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile]: good catch 👍

File: hazelcast-enterprise/Dockerfile
Comment (hasancelik) [reply hazelcast-enterprise/Dockerfile]: Actually, their(RedHat Registries) latest tag usage is a little bit different from DockerHub, `latest` tag is like symbolic link at linux. They just change its pointer, do not push separate image for it. But your points are valid, I have changed it to specific version(8.3) 👍

**Review by leszko — APPROVED** [review]

**Review by huseyinbabal — APPROVED** [review]
At security and operational level, you always discuss lots of this, my main criteria is this image should be carefully tested in k8s level. If integration team is already ok with this, LGTM!

**Pull request merged**
